include/Beman/Optional26/optional.hpp

```diff
@@ -229,719 +125,979 @@ inline constexpr auto format_kind<beman::optional26::optional<T>> = range_format
+    !is_convertible_v<optional<U>&&, T> && !is_convertible_v<const optional<U>&, T> &&
+    !is_convertible_v<const optional<U>&&, T> && !is_assignable_v<T&, optional<U>&> &&
+    !is_assignable_v<T&, optional<U>&&> && !is_assignable_v<T&, const optional<U>&> &&
+    !is_assignable_v<T&, const optional<U>&&>;
+} // namespace detail
 
+namespace detail {
 template <class T>
-class optional {
-    static_assert((!std::is_same_v<T, std::remove_cv_t<in_place_t>>) && (!std::is_same_v<std::remove_cv_t<T>, nullopt_t>));
-
-    struct empty {};
-    union {
-        empty _{};
-        T     value_;
-    };
-    bool engaged_ = false;
+    concept is_optional = requires(const T& t) { // exposition only
+    []<class U>(const optional<U>&) {}(t);
+};
+} // namespace detail
 
-    template <class... Args>
-    constexpr void construct(Args&&... args) {
-        std::construct_at(std::addressof(value_), std::forward<Args>(args)...);
-        engaged_ = true;
-    }
+// 22.5.3.1 General[optional.optional.general]
 
-    constexpr void hard_reset() noexcept {
-        value_.~T();
-        engaged_ = false;
-    }
+template <class T>
+class optional {
+    static_assert((!is_same_v<T, remove_cv_t<in_place_t>>) && (!is_same_v<remove_cv_t<T>, nullopt_t>));
+    static_assert(is_object_v<T> && !is_array_v<T>);
 
   public:
     using value_type = T;
     // Since P3168R2: Give std::optional Range Support.
-    using iterator       = detail::contiguous_iterator<T, optional>;       // see [optional.iterators]
-    using const_iterator = detail::contiguous_iterator<const T, optional>; // see [optional.iterators]
-
-    constexpr optional() noexcept
-        : _(), engaged_(false) {}
-
-    constexpr ~optional()
-        requires(!std::is_trivially_destructible_v<T>)
-    {
-        if (has_value())
-            value_.~T();
-    }
-
-    constexpr ~optional()
-        requires std::is_trivially_destructible_v<T>
-    = default;
-
-    constexpr optional(nullopt_t) noexcept {}
+    using iterator       = detail::contiguous_iterator<T, optional>; // see~\ref{optional.iterators}
+    using const_iterator = detail::contiguous_iterator<const T,
+                                                       optional>; // see~\ref{optional.iterators}
 
+    // \ref{optional.ctor}, constructors
+    constexpr optional() noexcept;
+    constexpr optional(nullopt_t) noexcept;
     constexpr optional(const optional& rhs)
-        requires std::is_copy_constructible_v<T> && (!std::is_trivially_copy_constructible_v<T>)
-    {
-        if (rhs.has_value()) {
-
-            std::construct_at(std::addressof(value_), rhs.value_);
-            engaged_ = true;
-        }
-    }
-
+        requires is_copy_constructible_v<T> && (!is_trivially_copy_constructible_v<T>);
     constexpr optional(const optional&)
-        requires std::is_copy_constructible_v<T> && std::is_trivially_copy_constructible_v<T>
+        requires is_copy_constructible_v<T> && is_trivially_copy_constructible_v<T>
```

Comment:
why do we prefer this convention ? using multiple using std::something at the top of the file